ginga/web/pgw/Widgets.py

```diff
@@ -1918,6 +2465,66 @@ def add_widget(self, child, row, col, stretch=0):
                              value=self.render_body())
         self.make_callback('widget-added', child)
 
+    def insert_cell(self, row, col):
+        indices = [row, col]
+        if self._rendered:
+            app = self.get_app()
+            app.do_operation("insert_cell", id=self.id, value=indices)
+
+    def insert_row(self, index):
+        indices = [index, self.num_cols]
+        self.num_rows += 1
+
+        # handle case where user inserts row at the end of the gridbox
+        if index == self.num_rows - 1:
+            for j in range(self.num_cols):
+                self.tbl[(index, j)] = Box()
+        else:
+            # shift key/value pairs down to make the row empty at index
+            for i in range(self.num_rows - 2, index - 1, -1):
```

Comment:
Also do this if you decide to move the counter increment to the end of the method.

```suggestion
            for i in range(self.num_rows - 1, index, -1):
```

💭 Although... is `dict` really the best data structure if you want to manipulate the indices like this? Why not Numpy array? Perhaps too late at this point, but I do wonder.